### DISCUSSION TITLE
File Write Fails on Solaris 10 (current main branch)

### DISCUSSION THREAD

**Discussion by Michael-Wohlstadter** [Netatalk/netatalk#629] [Q&A]
I am attempting to build the current 3.1 stream on Solaris 10 (sparc).  After a couple initial errors preventing the build, I can now build Netatalk.  Huge thanks to Mark for amazingly rapid response!  However, I am unable to copy any files to the mounted share.

As a couple reality checks, I build Netatalk 3.1.8 on the Solaris 10 box and I built Netatalk (current main branch) on Ubuntu 20.04.  In both cases I tried to keep the configurations as close as possible.  Both Solaris 10 / Netatalk 3.1.8 and Ubuntu 20.04 / Netatalk main build, install, and run as expected.  When I copied a file to the mounted share, it worked correctly in both of the test cases.

So I don't know if there is still an issue with the current branch on Solaris 10 or if I have a configuration issue.

For both Solaris 10 builds, I had previously installed appropriate libraries and tools from OpenCSW.  These include: autoconf, libtool, automake, gsed, pkgconfig, ggrep, libgcrypt, berkeley48devel, binutils, and libevent_dev.

**Solaris 10 / Netatalk main**
```
env PKG_CONFIG_PATH=/opt/csw/lib/pkgconfig ./configure \
	--with-init-style=solaris \
	--with-libgcrypt-dir=/opt/csw \
	--with-pam-confdir=/etc \
	--with-bdb=/opt/csw/bdb48 \
	--with-mysql-config=no
```
afp.conf
```
;
; Netatalk 3.x configuration file
;

[Global]
        hostname = Cortana
        log file = /var/log/netatalk.log
        log level = default:debug
        uam list = uams_dhx2.so
        mimic model = Xserve
        ea = auto
        zeroconf = no

[Source]
        path = /srv/net/data
        file perm = 0660
        directory perm = 2770
        umask = 0007
        valid users = @devusers
```

**Solaris 10 / Netatalk 3.1.8**
```
env PKG_CONFIG_PATH=/opt/csw/lib/pkgconfig ./configure \
	--with-init-style=solaris \
	--with-libgcrypt-dir=/opt/csw \
	--with-pam-confdir=/etc \
	--with-bdb=/opt/csw/bdb48 \
	--with-mysql-config=no
```
afp.conf
```
;
; Netatalk 3.x configuration file
;

[Global]
        hostname = Cortana
        log file = /var/log/netatalk.log
        log level = default:debug
        uam list = uams_dhx2.so
        mimic model = Xserve
        ea = auto
        zeroconf = no

[Source]
        path = /srv/net/data
        file perm = 0660
        directory perm = 2770
        umask = 0007
        valid users = @devusers
```

**Ubuntu 20.04 / Netatalk main**
```
./configure \
	--with-init-style=debian-systemd \
	--with-pam-confdir=/etc/pam.d \
	--with-mysql-config=no
```
afp.conf
```
;
; Netatalk 3.x configuration file
;

[Global]
        hostname = Alhazen
        log file = /var/log/netatalk.log
        log level = default:debug
        uam list = uams_dhx2.so
        mimic model = Xserve
        ea = auto
        zeroconf = no

[Model]
        path = /srv/net/data
        file perm = 0660
        directory perm = 2770
        umask = 0007
        valid users = @devusers
```

Thoughts, comments?  I'm not sure what to check next.  If it seems to be a bug in the current stream, I'll submit and issue.

Thanks!

Attached are the log files with level set to default:debug.
[netatalk-solaris-main.log](https://github.com/Netatalk/netatalk/files/13781920/netatalk-solaris-main.log)
[netatalk-solaris-3.1.8.log](https://github.com/Netatalk/netatalk/files/13781921/netatalk-solaris-3.1.8.log)
[netatalk-ubuntu-main.log](https://github.com/Netatalk/netatalk/files/13781922/netatalk-ubuntu-main.log)

**Comment by rdmark** [comment]
The first obvious thing to ask... are the `cnid_metad` and `cnid_dbd` daemons running? Without them you won't have access to the Berkeley DB to write file system data to.

FWIW I merged some improved error handling for this to the `main` branch a few hours ago. Please pull the latest sources if you haven't already done so.

**Comment by Michael-Wohlstadter** [comment]
There were two cnid processes running, cnid_met and cnid_dbd.  I re-pulled the source from main and built it.  I also ran dbd -f to rebuild the database.  I'm still having the same issue.

Note the permissions on the shared directory are 2770 and g+s.  The directory has the devusers group assigned and the login account is a member of that group.

Thanks,
Michael

**Comment by rdmark** [comment]
I can reproduce the same problem in my Solaris 10 VM. Something derails when it tries to write to the Berkeley DB I think, but the logging isn't entirely helpful (yet.)

In a few hours I'm taking off for a trip overseas, but I'll come back in a week or so to look deeper in what's going on here!

**Comment by Michael-Wohlstadter** [comment]
I'll setup with a newer version of Berkeley (5.3 compiles on my box) and try a few other things on my end.  You have have been wonderfully responsive and this is a home hobbyist setup.  Ergo no rush.  Enjoy your trip!

**Comment by rdmark** [comment]
@Michael-Wohlstadter Happy new year!

I took another stab at this today, but didn't really make any progress. I starting to suspect something to do with file system permissions / ACLs. I tried setting up a test volume, chmod 2775, then set `appledouble = v2` in afp.conf to force AppleDouble metadata (instead of file system Extended Attributes). With that volume, my user doesn't even have permissions to attempt to copy files to it, and in the logs I see errors where afpd tries to mkdir the various dot files that AppleDouble needs .... and gets permission denied all over the place.

If you search the mailing list archives there are countless threads with people trying to solve file system permissions issues on Solaris: https://sourceforge.net/p/netatalk/mailman/

There might be some configuration required in Solaris that I am not aware of!

**Reply by Michael-Wohlstadter** [reply]
Version 3.1.8 works without issue.  Is there a change between 3.1.8 and the current version that would require a change in ACLs in Solaris?  I didn't find anything in the mailing list that was helpful.

**Reply by rdmark** [reply]
@Michael-Wohlstadter Did you use the netatalk 3.1.8 package exclusively, or did you also try building it from source yourself?

There are three steps that I'd like to try:
- Validate that 3.1.8 compiled from scratch works (to eliminate any potential magic that happened during packaging.)
- Do a binary search through the release history to narrow down when Solaris 10 compatibility was ruined
- Do a binary search through the commit history leading up to that release version to narrow down the offending commit

I had a quick glance at the changelogs between 3.1.9 and 3.1.18 and we did touch upon the ACL code a few times. Either of those could have had a side effect.

FYI I recently compiled a list of release notes on the website for easy access: https://netatalk.sourceforge.io/documentation.php

**Reply by Michael-Wohlstadter** [reply]
Yes, I built from source.  Here are the statements:
```
git clone --depth 1 --branch netatalk-3-1-8 https://github.com/Netatalk/netatalk.git
cd netatalk
./bootstrap
env PKG_CONFIG_PATH=/opt/csw/lib/pkgconfig ./configure \
	--with-init-style=solaris \
	--disable-static \
	--enable-quota=no \
	--with-libgcrypt-dir=/opt/csw \
	--with-pam-confdir=/etc \
	--with-bdb=/opt/csw/bdb48 \
	--with-mysql-config=no

make
```

I will search next and get back to you.

Thanks!

**Reply by Michael-Wohlstadter** [reply]
Also, as I recall 3.1.9 is where the sys/attr.h issue began with Solaris 10.

**Reply by rdmark** [reply]
Ah yes, you're absolutely right. The fix for https://github.com/Netatalk/netatalk/issues/36 was what broke Solaris 10 compatibility in the first place, which is part of the 3.1.9 release. So either revert that patch every time you test, or cherry-pick my fix for the same. A bit of a pain, but it is what it is.

My patch for your convenience: https://github.com/Netatalk/netatalk/commit/91629e1ba5201ff06a6d73429c8d23bf15d94bec

**Reply by Michael-Wohlstadter** [reply]
The behavior is different but there is an issue introduced in 3.1.13 (3.1.12 runs fine).  The error is that I cannot connect with 3.1.13.  The error log (with max debug) reports "INTERNAL ERROR: Signal 11" after attempting to create the extended attribute metadata.

I have attached the log.

I will search through the individual commits for 3.1.13 next.

[debug.log.zip](https://github.com/Netatalk/netatalk/files/14105614/debug.log.zip)

**Reply by rdmark** [reply]
Good catch. In fact, 3.1.13 is hopelessly broken on any platform, when it encounters certain metadata. It was released without proper testing (tested only against a fresh volume without existing metadata).

This is the patch that fixed this crash, IINM: https://github.com/Netatalk/netatalk/pull/174

**Reply by Michael-Wohlstadter** [reply]
Compile on 3.1.14 fails on an undeclared LOCK_EX and LOCK_SH.  There is a note in the release notes for 3.1.15 that sys/file.h was included for LOCK_EX, but that define doesn't exist in the Solaris 10 version of sys/file.h.  (Note, 3.1.16 also fails on LOCK_EX and that's when I checked for LOCK_EX in the release notes.)

**Reply by rdmark** [reply]
It is not surprising that something broke in 3.1.14. It has a huge change log. In fact, this is where the Solaris attr patch originates that broke the compilation on Solaris 10 that you reported initially. 

The good news is that it was a fairly recent change that’s causing the current trouble.

**Reply by rdmark** [reply]
@Michael-Wohlstadter  were you able to get any closer to the commit that caused the breakage?

as a side note, I would be curious to know how well (or poorly) the Netatalk 2.3 branch works on your system. If it is SPARC architecture, the AppleTalk kernel module for Solaris/SPARC should build as well for the ultimate vintage server setup.

**Reply by Michael-Wohlstadter** [reply]
@rdmark sorry I got pulled into a project.  I should be able to dig back into it next week.  As to the 2.3 branch, I will try to build that and let you know.

**Reply by Michael-Wohlstadter** [reply]
@rdmark I'm not sure when I'll get back to working with Netatalk on Solaris.  My best guess is May.  Going through the individual patches will be a learning experience for me since my IT experience is mostly working with databases and scripting and not development.  Ergo while I can pull a release, setup the environment, and test the compilation, I really don't have a workflow for applying patches within a release.  I'm willing to dive in, but it will take a greater time commitment than I have.  Hopefully in May.

**Comment by Michael-Wohlstadter** [comment]
@rdmark I also didn't have much luck with the current development stream.  However, I did compile 3.1.8 and that works as expected.  While I'm not suggesting that ACLs aren't involved, it appears that there is at least some interaction with a change from 3.1.8 to the current version.

That said, I will look through the mailing list to see if I can't find something helpful.

Thanks,
Michael
